src/compiler/types.ts

```diff
@@ -4760,6 +4762,14 @@ export interface Program extends ScriptReferenceHost {
     /** @internal */ getAutomaticTypeDirectiveResolutions(): ModeAwareCache<ResolvedTypeReferenceDirectiveWithFailedLookupLocations>;
     isSourceFileFromExternalLibrary(file: SourceFile): boolean;
     isSourceFileDefaultLibrary(file: SourceFile): boolean;
+    /**
+     * Calculates the final resolution mode for a given module reference node. This is the resolution mode explicitly provided via import
+     * attributes, if present, or the syntax the usage would have if emitted to JavaScript. In `--module node16` or `nodenext`, this may
+     * depend on the file's `impliedNodeFormat`. In `--module preserve`, it depends only on the input syntax of the reference. In other
+     * `module` modes, when overriding import attributes are not provided, this function returns `undefined`, as the result would have no
+     * impact on module resolution, emit, or type checking.
+     */
+    getModeForUsageLocation(file: SourceFile, usage: StringLiteralLike): ResolutionMode;

```

Comment:
FYI @sheetalkamat:

I made this API public and suggested using it instead of `ts.getModeForUsageLocation`. I think we could consider deprecating that one.

However, when replacing all of our own usage of `getModeForUsageLocation` with `program.getModeForUsageLocation`, I noticed that it was almost _always_ used in conjunction with `program.getResolvedModule`. So, I made the internal API `program.getResolvedModuleFromModuleSpecifier` above to simplify. I think we could also consider making that one public, since getting the resolution of an import is extremely useful, and the new function does so without exposing the complexity of `ResolutionMode`.